src/box/box.cc

```diff
@@ -797,14 +823,139 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
+ * mixed transactions.
+ */
+static struct wal_row *
+wal_stream_save_row(struct wal_stream *stream,
+		    const struct xrow_header *row)
+{
+	static int64_t lsr_id = 0;
+	struct lsregion *lsr = &stream->lsr;
+	struct wal_row *new_row =
+		(struct wal_row *)xlsregion_alloc(lsr, sizeof(struct wal_row),
+						  ++lsr_id);
+	new_row->lsr_id = lsr_id;
+	new_row->row = *row;
+
+	assert(new_row->row.bodycnt <= 1);
+	if (new_row->row.bodycnt == 1) {
+		size_t len = new_row->row.body[0].iov_len;
+		char *new_base = (char *)xlsregion_alloc(lsr, len, lsr_id);
+		memcpy(new_base, new_row->row.body[0].iov_base, len);
+		/* Adjust row body pointers. */
+		new_row->row.body[0].iov_base = new_base;
+	}
+	return new_row;
+}
+
+/**
+ * Find the min lsr_id that is still needed.
+ */
+static int64_t
+wal_stream_find_min_lsr_id(struct wal_stream *stream)
+{
+	int64_t min_lsr_id = INT64_MAX;
+	struct rlist *nodes_rows = stream->nodes_rows;
+	struct wal_row *item;
+	/*
+	 * For each new row, lsr_id is incremented by 1. Only row from
+	 * different replica_id can get mixed, so for each replica_id,
+	 * the very first row has the smallest lsr_id of all row with that
+	 * replica_id. Thus, for each transaction read to the end, to
+	 * iterate through the nodes_rows array of lists and see which of
+	 * the first `row` for each `replica_id` has the lowest lsr_id.
+	 * This lsr_id is still needed, but anything less than this is not.
+	 * So can free everything up to this lsr_id.
+	 */
+	for (uint32_t i = 0; i < VCLOCK_MAX; i++) {
+		if (!rlist_empty(&nodes_rows[i])) {
+			item = rlist_first_entry(&nodes_rows[i], wal_row,
+						 in_row_list);
+			if (item->lsr_id <= min_lsr_id)
+				min_lsr_id = item->lsr_id - 1;
+		}
+	}
+	return min_lsr_id;
+}
+
+/**
+ * Deallocating memory for the wal_row
+ */
+static void
+wal_stream_gc(struct wal_stream *stream)
+{
+	struct lsregion *lsr = &stream->lsr;
+	int64_t lsr_id = wal_stream_find_min_lsr_id(stream);
+	lsregion_gc(lsr, lsr_id);
+}
+
+/**
+ * When restoring the log is read row-by-row. However, it is necessary
+ * to store the rows for correct further recovery. For example, with
+ * mixed transactions. The function saves the again coming string to the
+ * rlist. This rlist is stored in nodes_rows array in replica_id cell.
+ * As soon as a row arrives with the is_commit flag set, the corresponding
+ * transaction tries to apply. If an error occurs the transaction is
+ * rolled back. And we continue to work with the next transaction.
+ */
+static int
+wal_stream_apply_mixed_dml_row(struct wal_stream *stream,
+			       struct xrow_header *row)
+{
+	/*
+	 * A local transaction should not be written to nodes_rows with
+	 * id equal to 0. It should be written to nodes_rows with the id
+	 * of the node from which we are recovering. Since a local
+	 * transaction can only exist on the node from which we are
+	 * recovering.
+	 */
+	uint32_t id = row->replica_id ? row->replica_id : instance_id;
```

Comment:
What is this is an anonymous replica? Its `instance_id` is zero then. And it can write to replica-local spaces, hence creating mixed transactions.

Reply:
I create new issue #8746